### PR TITLE
Add support for a Python 3 build on RHEL 7

### LINKED ISSUE #27631: Red Hat 7: Build and package mantid against Python 3
Tasks:
- [x] Build remaining python36 packages RedHat
- [ ] Test build and packaging

### PR DESCRIPTION
**Pull request by martyngigg**
**Description of work.**

Contains packaging and other fixes when compiling against python 3 on RHEL 7. Note that this does not enable Python 3 by default yet. It includes building a private copy of sip/PyQt4 in Python 3 as the sip available from EPEL causes PyQt4 to segfault.

**To test:**

It currently requires the standard mantid/mantid copr and [my sandbox](https://copr.fedorainfracloud.org/coprs/martyngigg/sandbox/). Install these with

```
yum install yum-plugin-copr
yum copr enable mantid/mantid
yum copr enable martyngigg/sandbox
```

Build with cmake and `-DPYTHON_EXECUTABLE=/usr/bin/python3`. Check MantidPlot and workbench.

Fixes #27631

*This does not require release notes* because **Python 3 migration will be announced separately**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by peterfpeterson — CHANGES_REQUESTED** [review]
Current round of testing included installing the developer rpm and rebuilding mantidplot/mantidworkbench with python2. `python36-PyQt4-devel` should be removed from the rhel7 requirements.

**Comment by martyngigg** [comment]
`mantid-developer` in https://copr.fedorainfracloud.org/coprs/martyngigg/sandbox/ has been rebuilt to remove the PyQt4 dependencies for Python 3 on RHEL7. Latest test RPM is at [mantidunstable-4.2.20200115.1546-1.el7.x86_64.rpm](https://files.mantidproject.org/kits/testing/mantidunstable-4.2.20200115.1546-1.el7.x86_64.rpm)

**Comment by peterfpeterson** [comment]
On a machine with all of the developer dependencies installed, I managed to `yum install mantid42-4.2.0-1.el7.x86_64.rpm mantidunstable-4.2.20200115.1546-1.el7.x86_64.rpm` without issue.

**Comment by peterfpeterson** [comment]
Building the pyuic with qt4 appears to not have the python path quite correct:
```
[11/547] Generating /home/pf9/code/mantid/scripts/ErrorReporter/ui_errorreport.py
FAILED: /home/pf9/code/mantid/scripts/ErrorReporter/ui_errorreport.py 
cd /home/pf9/build/mantid-test/scripts/ErrorReporter && /usr/bin/python3 /home/pf9/build/mantid-test/extern-pyt4-sip/install/lib/site-packages/PyQt4/uic/pyuic.py /home/pf9/code/mantid/scripts/ErrorReporter/errorreport.ui -o /home/pf9/code/mantid/scripts/ErrorReporter/ui_errorreport.py && /usr/bin/python3 /home/pf9/code/mantid/buildconfig/wrap_pyui.py /home/pf9/code/mantid/scripts/ErrorReporter/ui_errorreport.py
Traceback (most recent call last):
  File "/home/pf9/build/mantid-test/extern-pyt4-sip/install/lib/site-packages/PyQt4/uic/pyuic.py", line 26, in <module>
    from PyQt4 import QtCore
ImportError: cannot import name 'QtCore'
```

**Comment by martyngigg** [comment]
@peterfpeterson I've added a tiny script to wrap pyuic and find the internal copy.

File: buildconfig/CMake/ExternalSipPyQt4.cmake [hunk 1]
```diff
@@ -51,21 +55,21 @@ ExternalProject_Add(extern-pyqt4
     --no-deprecated
     --qmake=/usr/bin/qmake-qt4
     --no-qsci-api
-  BUILD_COMMAND make 2> build.log
+  BUILD_COMMAND make -j24 2> build.log
```
Comment (peterfpeterson) [inline buildconfig/CMake/ExternalSipPyQt4.cmake]: You should use [ProcessorCount](https://cmake.org/cmake/help/v3.0/module/ProcessorCount.html) here

Comment (martyngigg) [reply buildconfig/CMake/ExternalSipPyQt4.cmake]: Thanks. I actually left that in by accident and meant to check how to do that with cmake.

Comment (martyngigg) [reply buildconfig/CMake/ExternalSipPyQt4.cmake]: Just pushed fix to use `ProcessorCount`.

**Comment by peterfpeterson** [comment]
It built!!!! For the testing:
```
$ bin/mantidpython 
.../build/mantid-test/bin/site.py:26: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp  # Avoid import loop in Python >= 3.3
/usr/lib/python3.6/site-packages/IPython/utils/traitlets.py:586: DeprecationWarning: inspect.getargspec() is deprecated since Python 3.0, use inspect.signature() or inspect.getfullargspec()
  argspec = inspect.getargspec(c)
/usr/lib/python3.6/site-packages/IPython/config/application.py:540: DeprecationWarning: generator 'Application._load_config_files' raised StopIteration
  for config in self._load_config_files(filename, path=path, log=self.log):
Python 3.6.8 (default, Jun 11 2019, 15:15:01) 
Type "copyright", "credits" or "license" for more information.

IPython 3.2.1 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.
```
which is a "normal" warning.

Successfully started `mantidplot`. Loaded some data and used the pick tab in instrument view to see the sub-window. `matplotlib.__version__` is 2.2.3.

Successfully started `mantidworkbench`. The warning from `mantidpython` appeared (didn't in `mantidplot`). In the messages window there are warnings:
```
.../code/mantid/qt/python/mantidqt/widgets/codeeditor/completion.py:74: DeprecationWarning: invalid escape sequence \(
  if re.search(builtin.__name__ + "\([\[\*a-zA-Z_].*\)", func_descriptor):
Welcome to Mantid 4.2.20200122.1309
Please cite: http://dx.doi.org/10.1016/j.nima.2014.07.029 and this release: http://dx.doi.org/10.5286/Software/Mantid
inspect.getargspec() is deprecated since Python 3.0, use inspect.signature() or inspect.getfullargspec()
invalid escape sequence \W
inspect.getargspec() is deprecated since Python 3.0, use inspect.signature() or inspect.getfullargspec()
inspect.getargspec() is deprecated since Python 3.0, use inspect.signature() or inspect.getfullargspec()
Not adding gui "ISIS_Reflectometry_Old.py"
Not adding gui "Frequency_Domain_Analysis_Old.py"
```
which can be fixed in a separate issue by supplying the regexp as a "raw" string with `r`. The argspec warning appears everytime something is executed in the ipython console which will get old fast.

**Review by peterfpeterson — APPROVED** [review]
Building with python3 on RHEL7 works as does `mantidpython`, `mantidworkbench`, and `mantidplot`.

**Comment by martyngigg** [comment]
@mantidproject/gatekeepers Another one if you wouldn't mind?

**Pull request merged**